File: jafu/src/main/java/org/springframework/fu/jafu/redis/AbstractRedisDsl.java
package org.springframework.fu.jafu.redis;

import org.springframework.boot.autoconfigure.data.redis.ClusterInitializer;
import org.springframework.boot.autoconfigure.data.redis.LettuceRedisInitializer;
import org.springframework.boot.autoconfigure.data.redis.RedisProperties;
import org.springframework.boot.autoconfigure.data.redis.SentinelInitializer;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.support.GenericApplicationContext;
import org.springframework.fu.jafu.AbstractDsl;

import java.time.Duration;
import java.util.List;
import java.util.function.Consumer;

import static java.util.Arrays.asList;

/**
 * base JaFu DSL for Redis configuration.
 *
 * @author Andreas Gebhardt
 */
public abstract class AbstractRedisDsl<SELF extends AbstractRedisDsl<SELF>> extends AbstractDsl {

    private final SELF self;

    private ApplicationContextInitializer<GenericApplicationContext> redisClientInitializer;

    private final Consumer<SELF> dsl;

    protected RedisProperties properties = new RedisProperties();

    protected abstract SELF getSelf();

    public void setRedisClientInitializer(ApplicationContextInitializer<GenericApplicationContext> redisClientInitializer) {
        this.redisClientInitializer = redisClientInitializer;
    }

    protected ApplicationContextInitializer<GenericApplicationContext> getRedisClientInitializier() {
        return redisClientInitializer;
    }

    public AbstractRedisDsl(final Consumer<SELF> dsl) {
        this.self = getSelf();
        this.redisClientInitializer = new LettuceRedisInitializer(properties);
        this.dsl = dsl;
    }

    public SELF database(final int database) {
        this.properties.setDatabase(database);
        return self;
    }

    public SELF url(final String url) {
        this.properties.setUrl(url);
        return self;
    }

    public SELF host(final String host) {
        this.properties.setHost(host);
        return self;
    }

    public SELF password(final String password) {
        this.properties.setPassword(password);
        return self;
    }

    public SELF port(final int port) {
        this.properties.setPort(port);
        return self;
    }

    public SELF ssl(final boolean ssl) {
        this.properties.setSsl(ssl);
        return self;
    }

    public SELF enableSsl() {
        this.properties.setSsl(true);
        return self;
    }

    public SELF timeout(final Duration timeout) {
        this.properties.setTimeout(timeout);
        return self;
    }

    public SELF clientName(final String clientName) {
        this.properties.setClientName(clientName);
        return self;
    }

    public SELF sentinel(final Consumer<SentinelDsl> sentinel) {
        this.properties.setSentinel(new RedisProperties.Sentinel());
        sentinel.accept(new SentinelDsl(this.properties.getSentinel()));
        return self;
    }

    public SELF cluster(final Consumer<ClusterDsl> cluster) {
        this.properties.setCluster(new RedisProperties.Cluster());
        cluster.accept(new ClusterDsl(this.properties.getCluster()));
        return self;
    }

    public SELF lettuce(final Consumer<LettuceDsl> lettuce) {
        setRedisClientInitializer(new LettuceRedisInitializer(properties));
        lettuce.accept(new LettuceDsl(this.properties.getLettuce()));
        return self;
    }

    public static class SentinelDsl {

        private final RedisProperties.Sentinel sentinel;

        SentinelDsl(final RedisProperties.Sentinel sentinel) {
            this.sentinel = sentinel;
        }

        public SentinelDsl master(final String master) {
            this.sentinel.setMaster(master);
            return this;
        }

        public SentinelDsl nodes(final List<String> nodes) {
            this.sentinel.setNodes(nodes);
            return this;
        }

        public SentinelDsl nodes(final String... nodes) {
            this.sentinel.setNodes(asList(nodes));
            return this;
        }

        public SentinelDsl password(final String password) {
            this.sentinel.setPassword(password);
            return this;
        }

    }

    public static class ClusterDsl {

        private final RedisProperties.Cluster cluster;

        ClusterDsl(final RedisProperties.Cluster cluster) {
            this.cluster = cluster;
        }

        public ClusterDsl nodes(final List<String> nodes) {
            this.cluster.setNodes(nodes);
            return this;
        }

        public ClusterDsl nodes(final String... nodes) {
            this.cluster.setNodes(asList(nodes));
            return this;
        }

        public ClusterDsl maxRedirects(final int maxRedirects) {
            this.cluster.setMaxRedirects(maxRedirects);
            return this;
        }

    }

    public static class PoolDsl {

        private final RedisProperties.Pool pool;

        PoolDsl(final RedisProperties.Pool pool) {
            this.pool = pool;
        }

        public PoolDsl maxIdle(final int maxIdle) {
            this.pool.setMaxIdle(maxIdle);
            return this;
        }

        public PoolDsl minIdle(final int minIdle) {
            this.pool.setMinIdle(minIdle);
            return this;
        }

        public PoolDsl maxActive(final int maxActive) {
            this.pool.setMaxActive(maxActive);
            return this;
        }

        public PoolDsl maxWait(final Duration maxWait) {
            this.pool.setMaxWait(maxWait);
            return this;
        }

        public PoolDsl timeBetweenEvictionRuns(final Duration timeBetweenEvictionRuns) {
            this.pool.setTimeBetweenEvictionRuns(timeBetweenEvictionRuns);
            return this;
        }

    }

    public static class LettuceDsl {

        private final RedisProperties.Lettuce lettuce;

        LettuceDsl(final RedisProperties.Lettuce lettuce) {
            this.lettuce = lettuce;
        }

        public LettuceDsl shutdownTimeout(final Duration shutdownTimeout) {
            this.lettuce.setShutdownTimeout(shutdownTimeout);
            return this;
        }

        public LettuceDsl pool(final Consumer<PoolDsl> pool) {
            this.lettuce.setPool(new RedisProperties.Pool());
            pool.accept(new PoolDsl(this.lettuce.getPool()));
            return this;
        }

        public LettuceDsl cluster(final Consumer<ClusterDsl> cluster) {
            cluster.accept(new ClusterDsl(this.lettuce.getCluster()));
            return this;
        }

        public static class ClusterDsl {

            private final RedisProperties.Lettuce.Cluster cluster;

            public ClusterDsl(final RedisProperties.Lettuce.Cluster cluster) {
                this.cluster = cluster;
            }

            public ClusterDsl refresh(final Consumer<RefreshDsl> refresh) {
                refresh.accept(new RefreshDsl(this.cluster.getRefresh()));
                return this;
            }

            public static class RefreshDsl {

                private final RedisProperties.Lettuce.Cluster.Refresh refresh;

                public RefreshDsl(final RedisProperties.Lettuce.Cluster.Refresh refresh) {
                    this.refresh = refresh;
                }

                public RefreshDsl period(final Duration period) {
                    this.refresh.setPeriod(period);
                    return this;
                }

                public RefreshDsl adaptive(final boolean adaptive) {
                    this.refresh.setAdaptive(adaptive);
                    return this;
                }

            }

        }
    }

    @Override
    public void initialize(GenericApplicationContext context) {
        super.initialize(context);

        dsl.accept(self);

        getRedisClientInitializier().initialize(context);
        new ClusterInitializer(properties.getCluster()).initialize(context);
        new SentinelInitializer(properties.getSentinel()).initialize(context);
    }

}

File: kofu-redis/src/main/kotlin/org/springframework/fu/kofu/redis/AbstractRedisDsl.kt
package org.springframework.fu.kofu.redis

import org.springframework.boot.autoconfigure.data.redis.RedisProperties
import org.springframework.boot.autoconfigure.data.redis.RedisProperties.*
import org.springframework.context.support.GenericApplicationContext
import org.springframework.fu.kofu.AbstractDsl
import java.time.Duration

/**
 * Kofu DSL base Redis DSL
 *
 * @author Waldemar Panas
 */
open class AbstractRedisDsl : AbstractDsl() {
	protected val properties = RedisProperties()

	/**
	 * Configure the database index used by the connection factory.
	 */
	var database: Int
		get() = properties.database
		set(value) {
			properties.database = value
		}

	/**
	 * Configure the connection URL. Overrides host, port, and password. User is ignored.
	 * Example: redis://user:password@example.com:6379
	 */
	var url: String?
		get() = properties.url
		set(value) {
			properties.url = value
		}

	/**
	 * Configure the redis server host.
	 */
	var host: String?
		get() = properties.host
		set(value) {
			properties.host = value
		}

	/**
	 * Configure the login password of the redis server.
	 */
	var password: String?
		get() = properties.password
		set(value) {
			properties.password = value
		}

	/**
	 * Configure the redis server port.
	 */
	var port: Int
		get() = properties.port
		set(value) {
			properties.port = value
		}

	/**
	 * Configure the connection timeout.
	 */
	var timeout: Duration?
		get() = properties.timeout
		set(value) {
			properties.timeout = value
		}

	/**
	 * Configure whether to enable SSL support.
	 */
	var ssl: Boolean
		get() = properties.isSsl
		set(value) {
			properties.isSsl = value
		}

	/**
	 * Configure the redis sentinel properties via a [dedicated DSL][SentinelDsl].
	 */
	fun sentinel(dsl: SentinelDsl.() -> Unit = {}) {
		properties.sentinel = Sentinel()
		SentinelDsl(properties.sentinel, dsl).initialize(context)
	}

	/**
	 * Configure the cluster properties via a [dedicated DSL][ClusterDsl].
	 */
	fun cluster(dsl: ClusterDsl.() -> Unit = {}) {
		properties.cluster = Cluster()
		ClusterDsl(properties.cluster, dsl).initialize(context)
	}
}

/**
 * Configure the cluster properties.
 *
 * @author Waldemar Panas
 */
class ClusterDsl(private val cluster: Cluster, private val init: ClusterDsl.() -> Unit) : AbstractDsl() {

	/**
	 * Configure the cluster node.
	 */
	fun node(host: String, port: Int) {
		if (cluster.nodes == null) {
			cluster.nodes = mutableListOf()
		}
		cluster.nodes.add("$host:$port")
	}

	/**
	 * Configure the maximum number of redirects to follow when executing commands
	 * across the cluster.
	 */
	var maxRedirects: Int?
		get() = cluster.maxRedirects
		set(value) {
			cluster.maxRedirects = value
		}

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		init()
	}
}

/**
 * Configure the redis sentinel properties.
 *
 * @author Waldemar Panas
 */
class SentinelDsl(private val sentinel: Sentinel, private val init: SentinelDsl.() -> Unit) : AbstractDsl() {

	/**
	 * Configure the name of the redis server.
	 */
	var master: String?
		get() = sentinel.master
		set(value) {
			sentinel.master = value
		}

	/**
	 * Configure the sentinel node.
	 */
	fun node(host: String, port: Int) {
		if (sentinel.nodes == null) {
			sentinel.nodes = mutableListOf()
		}
		sentinel.nodes.add("$host:$port")
	}

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		init()
	}
}

/**
 * Configure the pool properties.
 *
 * @author Waldemar Panas
 */
open class PoolDsl(private var pool: Pool, private val init: PoolDsl.() -> Unit) : AbstractDsl() {

	/**
	 * Configure the maximum number of "idle" connections in the pool. Use a negative value to
	 * indicate an unlimited number of idle connections.
	 */
	var maxIdle: Int
		get() = pool.maxIdle
		set(value) {
			pool.maxIdle = value
		}

	/**
	 * Configure the target for the minimum number of idle connections to maintain in the pool. This
	 * setting only has an effect if both it and time between eviction runs are
	 * positive.
	 */
	var minIdle: Int
		get() = pool.minIdle
		set(value) {
			pool.minIdle = value
		}

	/**
	 * Configure the maximum number of connections that can be allocated by the pool at a given
	 * time. Use a negative value for no limit.
	 */
	var maxActive: Int
		get() = pool.maxActive
		set(value) {
			pool.maxActive = value
		}

	/**
	 * Configure the maximum amount of time a connection allocation should block before throwing an
	 * exception when the pool is exhausted. Use a negative value to block
	 * indefinitely.
	 */
	var maxWait: Duration?
		get() = pool.maxWait
		set(value) {
			pool.maxWait = value
		}

	/**
	 * Configure the time between runs of the idle object evictor thread. When positive, the idle
	 * object evictor thread starts, otherwise no idle object eviction is performed.
	 */
	var timeBetweenEvictionRuns: Duration?
		get() = pool.timeBetweenEvictionRuns
		set(value) {
			pool.timeBetweenEvictionRuns = value
		}

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		pool = Pool()
		init()
	}
}
